### ISSUE TITLE
First-party tile maps

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#13782]
## What problem does this solve or what need does it fill?

Tile maps are a key part of many 2D games, both 2D and 3D.

There is a large amount of complexity here: API design, performance and features are all challenging.

## What solution would you like?

Bevy should make its own tilemap crate (one day) using a working group.

Some features:

- y-sorting
- LDTK and Tiled format imports

## What alternative(s) have you considered?

We could upstream one of the existing tilemap crates:

1. `bevy_ecs_tilemap`: challenging to maintain, complex API, poor performance at high map sizes. Lots of features though.
2. `bevy_tilemap`: dead since Bevy 0.5
3. `bevy_entitles`: a newer candidate, I'm personally unsure how this compares.

None of these strike me as exceptionally good candidates, but the working group should look closely at this for prior art.

## Additional context

This is one of the prime formats for content authoring in the editor. Without a first-party tilemap format, the editor will be much less useful for 2D games.

This is a large initiative, please don't try to tackle this without first forming a working group.

**Comment by IceSentry** [comment]
I personally think bevy should first have a first party way to render a tilemap. Without any fancy features like LDTK or Tiled integration, once we have a first party tilemap rendering solution it will be way easier for third party to support those integrations. 

I'm also not sure if y-sorting should be part of a tilemap render. The way I see it, tiles are intended to be behind everything and if you want to add spares entities with y-sorting that should be done with mesh2d/sprites

**Comment by alice-i-cecile** [comment]
I'm fully on board with that as an initial solution :) Rendering and a standard performant data structure are the hardest bits IMO.

**Comment by MScottMcBee** [comment]
In my game, most entities live on a layer in between background and foreground tilemap layers. I don't think tiles need their own sorting, but handling multiple layers with the ability to control layer sorting is important.

**Comment by IceSentry** [comment]
> I disagree. Y-sorting is pretty much a requirement for isometric tilesets.

It doesn't need to be part of the renderer. All the renderer needs is to take depth into account even in 2d. I already have a PR for that. Once you have depth you can just set it based on the z-index and trivially implement y-sorting user side. The renderer doesn't need to know about y-sorting.

**Comment by NoahShomette** [comment]
> Alternatives
Another alternative that is more similar to bevy_tilemap than any other current crate is my crate [bevy_sparse_tilemap](https://github.com/NoahShomette/bevy_sparse_tilemap). It's pretty simple feature wise and a bit clunky to look at but it's pretty simple to use and interact with and has very good baseline performance due to dodging the high entity count issue that the other crates run into.

The main goal for this crate was to *not* do an entity for every tile. This crate supports both dense and sparse maps and dense is basically a `Vec<Vec<TileData>>`. It has a built in escape hatch to spawn/map an entity to a given tile so you can keep track of unique data/tiles that aren't default anymore. This keeps entity counts very low which really impacts performance rn. This crate also doesn't currently handle any rendering (I use bevy_fast_tilemap in the projects I have that use this crate).

The crate currently supports multiple tilemap types, very very large maps (15k x 15k in my tests), and I have plans to add more helper related stuff like tile->world, iteration functions, as well as I'd like to fix the generics hell it's in rn from the refactor to support multiple tilemap types.

**Comment by jamescarterbell** [comment]
Tilemaps are a bit cursed in bevy right now for two reasons:

1. Deciding whether to represent tiles as entities is a big decision: it seems reasonable on its face but doesn't seem to scale well for a variety of reasons.
2. Without real secondary indexing support, accessing tiles is a PITA and keeping any secondary index up to date is also a pain.

On top of that, any built in tilemaps would have to deal with chunking or not chunking, infinite vs not infinite, tilemaps themselves as entities, etc...

For what it's worth, I also have a crate that's just dealing with tile accessing APIs, but I have some new ideas to make maintaining the indices much nicer:

https://github.com/jamescarterbell/bevy_tiles

(I'll freely give up the name btw)

**Comment by clarfonthey** [comment]
Just poking in here as I look for solutions for myself, but honestly, a good "first pass" version of this would just be to cover the rendering step, since it feels like the representation for rendering vs. accessing the tiles will probably be different, and cases where the tiles just correspond to background decorations are common enough that people probably won't want to access the tiles at all outside of rendering.

Likely something that can take a texture, metrics to get tiles from that texture, and metrics to render tiles onto a grid, then expose an API for rendering chunks based upon decisions about culling, etc. Even if you want to do something more complicated like choosing tiles depending on terrain or a bitmask, it's ultimately going to fall onto the renderer to take that information and render the tiles individually, and I don't think there'd ever be a situation where the decided-upon tiles would be more data to send than the data that decided them, so, it's probably worth just doing the least amount of work on the GPU here.

The main issue is designing an API to lazy-load the chunks as they're needed, allowing some other system to populate them. That system could be part of a larger "nice" loader that would include the other expected features of a tilemap system, but doesn't have to be fixed to any one rendering strategy.

At least to me, the main benefit of a tilemap system is the ability to render a large amount of tiles without having to manually enumerate them. I tell it a layer of tiles, maybe in a sparse or dense mode, and it just renders all of them without a fuss. Whether those tiles are squares, hexagons, rhombi, triangles, or some secret other thing doesn't really matter to the rendering engine, since it's probably just going to find a way to draw a bunch of rectangles from a texture as efficiently as possible.

**Comment by ChristopherBiscardi** [comment]
> poor performance at high map sizes

@alice-i-cecile is this referencing the [Transform propagation issue](https://github.com/bevyengine/bevy/pull/14384) that got fixed this cycle, or is it in reference to something else?

**Comment by alice-i-cecile** [comment]
That was the primary cause, along with rendering world extraction. I'd love more up-to-date benchmarks / evaluation!

**Comment by StarArawn** [comment]
> bevy_ecs_tilemap: challenging to maintain, complex API, poor performance at high map sizes. Lots of features though.
Poor performance was due to bevy transform propagation issues, although I haven't checked to see if this is fixed in more recent bevy versions. I don't have the time to try and solo rewrite `bevy_ecs_tilemap` to be bevy first, but I'm happy to work with someone to accomplish this.

**Comment by ChristopherBiscardi** [comment]
transform propagation was [improved in 0.15](https://github.com/bevyengine/bevy/pull/14384) and in 0.16 with [parallel propagation](https://github.com/bevyengine/bevy/pull/17840) and [static subtrees](https://github.com/bevyengine/bevy/pull/18589) as well. I was planning on working in ecs_tilemap to bring it closer to upstreaming but it didn't seem like that was the desired approach by upstream so I didn't continue working on it. Currently planning to migrate ecs_tilemap on top of whatever gets merged for feedback purposes, etc.

My understanding is that a from-scratch minimal-feature approach that doesn't consider 3rd party crates like [18866](https://github.com/bevyengine/bevy/pull/18866) is desired at the moment.

**Comment by cpetzold** [comment]
I'd be excited to start discussing this as a working group, but I know there's some worry about stifling momentum on getting building blocks merged first. The significant number of 3rd party tilemap crates kind of shows that it's a hard-to-generalize problem, so I could see it being a bit of a rabbit hole. I do think that based on the latest improvements to transform propagation, working towards upstreaming and improving on ecs_tilemap makes the most sense to me. I wonder if it would be possible/difficult to adapt ecs_tilemap to use Mesh2d/Material2d directly, since that seems desirable and could potentially simplify the code?

**Comment by clarfonthey** [comment]
As I mentioned, I think that starting with the rendering step is a good idea since it's the most important to optimise, and while the APIs to create the actual tiles and manipulate them may vary wildly, all of them will all ultimately rely on drawing a fixed grid of tiles with fixed offfets, even if those offsets and the tiles can vary wildly in shape and might be sourced from different assets.

I think that every solution will probably have some separate representation between the rendering API and the logical tile API, at least a little bit, based upon individual desires and purposes. But the rendering can probably exist on its own regardless of that.

For example, I think that one entity per tile is dramatic overkill for a lot of situations, but it importantly isn't for all situations. And in both cases, you'll need some kind of unified system and shader for rendering the tiles which is optimised to run fast. I think having something that can:

* Track the texture with the tiles
* Track what tiles need to be drawn, and in what pattern (square, triangular, hexagonal, etc. are all just differing grid offsets)h
* Be able to dynamically load in chunks of tiles that will be drawn later but aren't drawn now

Is a good starting point that will do a lot of the heavy lifting, such that people can still build their own APIs on top of that before a central one is decided upon.

**Comment by ConnerPetzold** [comment]
Yeah that's fair, something like that is what we've been working towards and have started with the super basic chunk rendering in #18866. It currently only handles square tiles without offsets, and doesn't have a higher-level API for laying the chunks out. But it should be easy enough to build on. I also agree that we can separate out rendering from the user-land API and be able to support entity-per-tile or not. Although I do think for the vast majority of cases entity-per-tile works and aligns well with the rest of Bevy, so it feels like maybe it should be opt-out.

The main reason I was suggesting working with ecs_tilemap is because it seems like a good base for making the improvements you've mentioned (separating rendering and logical API's, opt-out entity-per-tile, etc.). I think it'd also  be fine to start from scratch and draw inspiration from ecs_tilemap and other tilemap crates.

**Comment by jamescarterbell** [comment]
"Although I do think for the vast majority of cases entity-per-tile works and aligns well with the rest of Bevy, so it feels like maybe it should be opt-out."

This isn't really true, it's good for the subset of cases where you need per tile information and interaction, but if you're just using tiles to build a level of simple geometry that is merged into one collider? You don't need it there, and that's a pretty big use case for tile maps.

I think having the ability to attach entities to tiles or do full entity per tile is good as an opt in, but usually not a good default.

**Comment by clarfonthey** [comment]
Yeah, I agree that in general you're going to want to cull or coalesce at least some of the tile entities for various performance or logic reasons, so, I don't think that assigning an individual entity per tile is a good idea in most cases. That's why I mentioned that focusing on the graphical aspect is more likely a big win, since the way tiles will connect to entities is pretty use-case-specific, but the way tiles are drawn probably won't be.

Will still want the proper tools for offsetting tiles and grids to support non-square layouts (isometric, hexagonal, etc.) but that mostly just ends up as some extra maths in the shader which is easily able to handle all the cases. Whereas I don't think that there's an easy answer to offering a generic system that links the tiles to entities, allows all the desired culling/coalescing logic, has the performance people want out of it, and is easier than coding your own solution.

**Comment by ConnerPetzold** [comment]
@jamescarterbell @clarfonthey there was some discussion in the working group about tile entities here: https://discord.com/channels/691052431525675048/1367261228459884604/1367289525642067988

The current direction based on that discussion and other explorations is to start with entity per tile, with the plan to follow up with a separate way to do render-only tilemap layers. The two approaches would use the same underlying recently merged TilemapChunk API, but would use separate tile stores.

**Comment by jamescarterbell** [comment]
>  start with entity per tile

I think this is the wrong place to start, we can see from this thread multiple people saying a tile per entity is overkill or doesn't fit their use case, so having that be the default is not the right choice for the first party solution.

I think a better alternative is to have a more generic `ChunkStorage<T>` and associated API's to make indexing into it from ecs systems easy, then to have entity per tile be modeled on top of that via `ChunkStorage<Entity>`.  This would already match the current approach for the rendering component of not being tile per entity, and would give people who want to opt out of that behavior a good model to follow.

Long term, I think we should have a setup like this:
TileMap containing map of chunks
Chunks containing ChunkStorages of various types
Entities with a TileOf(EntityWithTileMap) psuedo relation (since it would end up a child of a chunk imo) and a TilePosition component

This would support people who want to have an ecs centric tile solution and those who really just need the chunk representation for tasks similar to rendering, such as collision shape generation for chunks.

I worked on a crate that is modeled like this last yearish, and I think it's a good model for all parties, here is a good example showing indexing entities by tile position using custom queries: https://github.com/jamescarterbell/bevy_tiles/blob/bevy_0_17/crates/bevy_tiles_ecs/examples/spatial.rs 

```

fn add_damage(
    mut commands: Commands,
    // Nested query of TileMap -> Chunk -> Query<D, F>
    mut block_maps: TileEntityMapQuery<(Entity, Option<&mut Damage>), With<Block>>,
    map: Query<(Entity, &TileMap, &TileDims, &TileSpacing), With<GameLayer>>,
    windows: Query<&Window, With<PrimaryWindow>>,
    camera: Query<(&Camera, &GlobalTransform)>,
    buttons: Res<ButtonInput<MouseButton>>,
) {
    let (map_id, map, dims, spacing) = map.single().unwrap();
    let (cam, cam_t) = camera.single().unwrap();
    // Get the map we care about
    let mut blocks = block_maps.get_map_mut(map_id).unwrap();

    // Get the tile positions we care about
    let cursor_pos = windows
        .single()
        .unwrap()
        .cursor_position()
        .and_then(|cursor| cam.viewport_to_world(cam_t, cursor.xy()).ok())
        .map(|ray| ray.origin.truncate())
        .map(|pos| world_to_tile(pos, *dims, Some(*spacing)));

    if let Some(damage_pos) = buttons
        .just_pressed(MouseButton::Left)
        .then_some(cursor_pos)
        .flatten()
    {
        let start = [damage_pos[0] - 2, damage_pos[1] - 2];
        let end = [damage_pos[0] + 2, damage_pos[1] + 2];
        // Use our magic nested query to figure out which blocks were damaged,
        for (block_id, damage) in blocks.iter_in_mut(start, end) {
            if let Some(mut damage) = damage {
                **damage += 1;
            } else {
                // Add damage to any block entities that haven't been damaged yet
                commands.entity(block_id).insert(Damage(1));
            }
        }
    }
}
```

Please feel free to poke around the crate, it's very command centric, too generic with 3d stuff, and not making use of relations now that those are stable and performant, also there's some iffy unsafe code in the iterators, but I think it does a relatively good job of making a bevy ecs centric API without necessarily having a tile per entity, unless you explicitly wanted.  As an example, here's the current bevy main example rewritten using the new query types and focus on chunk storage (Explicitly reusing the awesome chunk renderer in bevy main, which I think was a great first step): https://github.com/jamescarterbell/bevy_tiles/blob/bevy_0_17/crates/bevy_tiles/examples/render.rs
